[SUREFIRE-964] TEST-*.xml files generated by Surefire throw validation warnings in Eclipse for no grammer constraints (DTD or XML schema) referenced in the document

lamyaa opened the pull request:
This pull request adds an XSD schema for the surefire XML reports.
I added a link to the XSD in `index.apt.vm`.

lamyaa commented:
@Tibor17 let me know what you think of this XSD

Tibor17 commented:
Can you please apply regex on time attribute?
This is my test of the format:
```
NumberFormat.getInstance( Locale.ENGLISH ).format( 1001.5134d ).matches( "(([0-9]{0,3},)*[0-9]{3}|[0-9]{0,3})\\.[0-9]{0,3}" )
NumberFormat.getInstance( Locale.ENGLISH ).format( 1.5134d ).matches( "(([0-9]{0,3},)*[0-9]{3}|[0-9]{0,3})\\.[0-9]{0,3}" )
```
You have "skipped" as an attribute. Thats ok, but the element testcase may contain elements in it with name "skipped". Can you please therefore check in class `SurefireReportGenerator` if we are missing some more?
If you have some spare time it would be fantastic to update `SurefireReportGenerator` by adding the schema location. If you need a help with DOXIA, Herve on our mailing list may help you with adding the schema header. Maybe he has a good example.

Tibor17 commented:
We need to check if all attributes are optional.
The most important would be to validate some report before going to release.

lamyaa commented:
@Tibor17 I added regex for the time attribute and added successrate which I had missed before. I also tried validating a couple of reports against this xsd on http://www.freeformatter.com and http://xmlvalidation.com. They seem fine.
I am not quite sure how to add the schema location in SurefireReportGenerator so I'll get in touch with Herve to get some help.

Tibor17 commented:
@lamyaa
Sorry I wrote wrong class name. It report is correctly generated by class `StatelessXmlReporter`.

hboutemy commented:
I suppose you expect your xsd to be copied to the generated html site: putting a .xsd file in apt directory won't do the job
see http://maven.apache.org/plugins/maven-site-plugin/examples/creating-content.html#Adding_Extra_Resources instead

hboutemy commented:
another idea: instead of handwriting the xsd, did you try to write a Modello model, and let Modello generate the xsd and a html documentation? That's the documentation part that is the most interesting IMHO wrt pure handwritten xsd (given the model seems quite simple)

Tibor17 commented:
Modello looks interesting 
https://maven.apache.org/guides/mini/guide-using-modello.html
Maybe JAXB is simple as well.
I guess surefire's class 'StatelessXmlReporter' would cost us some time to adapt to Modello in order to understand Java sources generated by Modello, however it could be nice refactoring in the future.
It could be finally object-oriented xml serialization, unlike it is now when we use hardcoded strings of xml elements `"testcase"`.

hboutemy commented:
you don't need to refactor everything to use full Modello, ie java generated model + XML serializer
even without using these, using Modello can give you a nice doc, like http://maven.apache.org/ref/current/maven-model/maven.html

currently, Modello dcumentation is not available, since codehaus shut down and nobody did the move to github, but if you're interested, I can work on getting Modello back to life (that has to be done sometime...)

Tibor17 commented:
@hboutemy 
yes please try to backport the documentation if you have some spare time. Thx.

lamyaa commented:
@hboutemy Thanks for your comments. I moved the xsd to resources and made sure that the link works. I don't know if we'll look at Modello any time soon but I can try to take a look if I get some documentation.

lamyaa commented:
@Tibor17 StatelessXmlReporter makes more sense :). I added the schema location there. Note that I updated the regex because I ran into a case where some test classes were running so fast that their time was "0". You may consider updating your test as well. Let me know if anything else needs to be done. Also, should I rebase the branch?

Tibor17 commented:
@lamyaa 
LGTM
No need to rebase the branch because can be merged with master.
I will run the buid.

Tibor17 commented:
@lamyaa 

Pls add these elements and attributes:
skipped, message, system-out, error.
These are my XML examples:
```
  <testcase name="pkg.BTest" classname="pkg.BTest" time="0">
    <skipped message="org.junit.runner.notification.StoppedByUserException"/>
    <system-out><![CDATA[$$$ RUN_CLASS]]></system-out>
  </testcase>
```

```
  <testcase name="someMethod" classname="pkg.ATest" time="0.009">
    <error message="assert &quot;foo&quot; == &quot;bar&quot;&#10;             |&#10;             false" type="java.lang.RuntimeException">java.lang.RuntimeException: 
assert "foo" == "bar"
             |
             false
	at pkg.ATest.someMethod(ATest.java:74)
</error>
    <system-out><![CDATA[pkg.ATest]]></system-out>
  </testcase>
```

Tibor17 commented:
Pls add optional "group" attribute in testsuite and testcase element.
Also system-err and error and failure elements within testcase element. The failure and error have attributes message and type. The Javadoc in StatelessXmlReporter class has good examples with these.

lamyaa commented:
@Tibor17 I looked at the javadoc in StatelessXmlReporter and added the missing elements and attributes. Do you think we should make the maxOccurs more specific for the elements inside testcase? I suppose there can only be one error/failure/ but many rerunFailure.
Also, for the non-optional attributes, should specifically mark them as required (with use="required")?

Tibor17 commented:
>> Do you think we should make the maxOccurs more specific for the elements
inside testcase? I suppose there can only be one error/failure/ but many
rerunFailure.

Somebody reported a bug in JIRA that ErrorCollection in JUnit reported more
than one failure per test method. I think this is alright as well as
collection is about many failures.
Regarding errors I guess this is max 1 as soon as the exception different
from AssertionError can be thrown just once. Regarding minOccurs it should
be 0.
Regarding rerunFailure this can be many or no occurs.
Regarding  system-out and system-err minOccurs=0 maxOccurs=1.


>>Also, for the non-optional attributes, should specifically mark them as
required (with use="required")?
yes please.

btw, I am going for holidays. I will be back after 12 days.
Then I will send a list of new candidates of committers including you to
ASF and we will proceed with your other poll requests.

On Thu, Aug 6, 2015 at 5:35 AM, lamyaa <notifications@github.com> wrote:

> @Tibor17 <https://github.com/Tibor17> I looked at the javadoc in
> StatelessXmlReporter and added the missing elements and attributes. Do you
> think we should make the maxOccurs more specific for the elements inside
> testcase? I suppose there can only be one error/failure/ but many
> rerunFailure.
> Also, for the non-optional attributes, should specifically mark them as
> required (with use="required")?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/maven-surefire/pull/99#issuecomment-128226143>.
>



-- 
Cheers
Tibor

Tibor17 commented:
I am open to upgrade surefire with Modello in version 3.1. The reason is that I am busy with 2.19 and 3.0 at the moment. Meanwhile JIRA should be reported to remember future activities.

Tibor17 commented:
The build does not pass because the file `surefire-test-report.xsd` is missing the license.
Feel free to use the license from whatever POM.xml as a XML comment starting on second line. Thx

lamyaa commented:
Done, sorry about that.

Tibor17 commented:
LGTM

Tibor17 commented:
This was merged with master. I am wondering why GitHub did not close this PR.

Tibor17 commented:
@lamyaa 
Pls close this pr. Thx
